FWCore/Framework/src/ProductResolvers.h

```diff
@@ -336,8 +331,17 @@ namespace edm {
                         ServiceToken const& token,
                         SharedResourcesAcquirer* sra,
                         ModuleCallingContext const* mcc) const final;
+    void putProduct_(std::unique_ptr<WrapperBase> edp) const final;
     bool unscheduledWasNotRun_() const final { return false; }
     bool productUnavailable_() const final;
+    void resetProductData_(bool deleteEarly) final;
+
+    constexpr static const ProductStatus defaultStatus_ = ProductStatus::NotPut;
+
+    // for filter in a Path
+    // The variable is only modified or read at times where the
+    //  framework has guaranteed synchronization between write and read
+    CMS_THREAD_SAFE mutable ProductStatus status_;
```

Comment:
Contrary to my earlier thought (https://github.com/cms-sw/cmssw/pull/28667#discussion_r360576272), I believe now that this variable could stay non-atomic. The `status_` is read in `resolveProduct_()` and `putProduct_()`, and written in `putProduct_()` and `resetProductData_()`. It is not used to synchronize anything else (via memory fence). If we indeed guarantee that `resolveProduct_()` and `putProduct_()` may not be called concurrently, it should be fine as it is.